benchmarks/low_level.cpp

```diff
@@ -0,0 +1,455 @@
+    for(auto _ : state) {
+        state.PauseTiming();
+        m_os.reset();
+        m_os << cp_tx;
+        m_of.close();
+        m_if.open(m_BENCHMARK_File);
+
+        read_cp = cbdc::transaction::compact_tx(read_tx);
+        m_is.reset();
+        state.ResumeTiming();
+        m_is >> read_cp;
+        state.PauseTiming();
+        ASSERT_EQ(cp_tx, read_cp);
+    }
+
+    m_if.close();
+}
+
+// roundtrip benchmarking full tx
+BENCHMARK_F(low_level, roundtrip_tx)(benchmark::State& state) {
+    // serialize 1-1 tx
+    auto read_tx = cbdc::transaction::full_tx();
+    m_valid_tx = wallet1.send_to(100, wallet2.generate_key(), true).value();
+
+    for(auto _ : state) {
+        state.PauseTiming();
+        m_of.open(m_BENCHMARK_File);
+
+        m_os.reset();
+        state.ResumeTiming();
+        m_os << m_valid_tx;
+        state.PauseTiming();
+        m_of.close();
+
+        read_tx = cbdc::transaction::full_tx();
+
+        m_if.open(m_BENCHMARK_File);
+        m_is.reset();
+        state.ResumeTiming();
+        m_is >> read_tx;
+        state.PauseTiming();
+        m_if.close();
+        ASSERT_EQ(read_tx, m_valid_tx);
+    }
+}
+
+// roundtrip benchmarking compact tx
+BENCHMARK_F(low_level, roundtrip_compact_tx)(benchmark::State& state) {
+    // serialize 1-1 tx
+    auto read_tx = cbdc::transaction::full_tx();
+    auto read_compact = cbdc::transaction::compact_tx(read_tx);
+    m_valid_tx = wallet1.send_to(100, wallet2.generate_key(), true).value();
+    auto cp_tx = cbdc::transaction::compact_tx(m_valid_tx);
+
+    for(auto _ : state) {
+        state.PauseTiming();
+        read_compact = cbdc::transaction::compact_tx(read_tx);
+
+        m_of.open(m_BENCHMARK_File);
+        m_os.reset();
+        state.ResumeTiming();
+        m_os << cp_tx;
+        state.PauseTiming();
+        m_of.close();
+
+        m_if.open(m_BENCHMARK_File);
+        m_is.reset();
+
+        state.ResumeTiming();
+        m_is >> read_compact;
+        state.PauseTiming();
+
+        m_if.close();
+        ASSERT_EQ(read_compact, cp_tx);
+    }
+}
+
+// wallet sign tx
+BENCHMARK_F(low_level, sign_tx)(benchmark::State& state) {
+    // sign 1-1 tx
+    for(auto _ : state) {
+        state.PauseTiming();
+        m_valid_tx
+            = wallet1.send_to(100, wallet2.generate_key(), true).value();
+
+        state.ResumeTiming();
+        wallet1.sign(m_valid_tx);
+        state.PauseTiming();
+
+        // send back to use again
+        wallet1.confirm_transaction(m_valid_tx);
+        wallet2.confirm_transaction(m_valid_tx);
+        m_valid_tx
+            = wallet2.send_to(100, wallet1.generate_key(), true).value();
+        wallet1.confirm_transaction(m_valid_tx);
+        wallet2.confirm_transaction(m_valid_tx);
+    }
+}
+
+// tx validiation
+BENCHMARK_F(low_level, valid_tx)(benchmark::State& state) {
+    // validate 1-1 tx
+    auto err = cbdc::transaction::validation::check_tx(m_valid_tx);
```

Comment:
So, this is validating (and below, failing validation for) the same tx over and over again. Is there a significant difference in-numbers if we validate a different transaction every time? If so, would the varied benchmark be more fair?

Reply:
This is a good question, and something I was wondering myself -- I'd agree that it's likely better to generate new transactions rather than reusing the same one, will make that change

Reply:
Quick note -- for the failed validation, I would say that the usage of the same transaction in this particular case is sensible, since it will continuously fail on an early step (before much processing occurs). However, this would suggest that different types of verification failures should be benchmarked as well.

Reply:
That's a reasonable point. I'd suggest marking it as an early-failout then, and maybe adding a todo (or we can open an issue after this is merged) to expand the failing benchmarks in the future.